fiona/_vsiopener.pyx

```diff
@@ -0,0 +1,491 @@
+    try:
+        file_opener = registry[key]
+    except KeyError as err:
+        errmsg = f"Opener not found: {repr(err)}".encode("utf-8")
+        CPLError(CE_Failure, <CPLErrorNum>4, <const char *>"%s", <const char *>errmsg)
+        return NULL
+
+    cdef object file_obj
+
+    try:
+        file_obj = file_opener.open(urlpath, mode)
+    except ValueError as err:
+        # ZipFile.open doesn't accept binary modes like "rb" and will
+        # raise ValueError if given one. We strip the mode in this case.
+        try:
+            file_obj = file_opener.open(urlpath, mode.rstrip("b"))
+        except Exception as err:
+            return NULL
+    except Exception as err:
+        return NULL
+
+    log.debug("Opened file object: file_obj=%r, mode=%r", file_obj, mode)
+
+    # Before we return, we attempt to enter the file object's context
+    # and store an exit callback stack for it.
+    stack = contextlib.ExitStack()
+
+    try:
+        file_obj = stack.enter_context(file_obj)
+    except (AttributeError, TypeError) as err:
+        log.error("File object is not a context manager: file_obj=%r", file_obj)
+        errmsg = f"Opener failed to open file with arguments ({repr(urlpath)}, {repr(mode)}): {repr(err)}".encode("utf-8")
+        CPLError(CE_Failure, <CPLErrorNum>4, <const char *>"%s", <const char *>errmsg)
+        return NULL
+    except FileNotFoundError as err:
+        errmsg = "OpenFile didn't resolve".encode("utf-8")
+        return NULL
+    else:
+        exit_stacks = _OPEN_FILE_EXIT_STACKS.get()
+        exit_stacks[file_obj] = stack
+        _OPEN_FILE_EXIT_STACKS.set(exit_stacks)
+        log.debug("Returning: file_obj=%r", file_obj)
+        return <void *>file_obj
+
+
+cdef int pyopener_eof(void *pFile) with gil:
+    cdef object file_obj = <object>pFile
+    if file_obj.read(1):
+        file_obj.seek(-1, 1)
+        return 1
+    else:
+        return 0
+
+cdef vsi_l_offset pyopener_tell(void *pFile) with gil:
+    cdef object file_obj = <object>pFile
+    return <vsi_l_offset>file_obj.tell()
+
+
+cdef int pyopener_seek(void *pFile, vsi_l_offset nOffset, int nWhence) with gil:
+    cdef object file_obj = <object>pFile
+    # TODO: Add "seekable" check?
+    file_obj.seek(nOffset, nWhence)
+    return 0
+
+
+cdef size_t pyopener_read(void *pFile, void *pBuffer, size_t nSize, size_t nCount) with gil:
+    cdef object file_obj = <object>pFile
+    cdef bytes python_data = file_obj.read(nSize * nCount)
+    cdef int num_bytes = len(python_data)
+    # NOTE: We have to cast to char* first, otherwise Cython doesn't do the conversion properly
+    memcpy(pBuffer, <void*><char*>python_data, num_bytes)
+    return <size_t>(num_bytes / nSize)
+
+
+cdef size_t pyopener_write(void *pFile, void *pBuffer, size_t nSize, size_t nCount) with gil:
+    cdef object file_obj = <object>pFile
+    buffer_len = nSize * nCount
+    cdef np.uint8_t [:] buff_view = <np.uint8_t[:buffer_len]>pBuffer
```

Comment:
As far as I can see, this is the only use of numpy in this PR. I _think_ this could be easily avoided by using a builtin cython type instead of a numpy type. Something like `unsigned char` instead of `np.uint8_t`

That would also avoid having to add numpy as a build dependency to start with?

Reply:
@jorisvandenbossche yes, I'll definitely look into removing the hard numpy requirement before the next pre-release.